comictaggerlib/ui/talkeruigenerator.py

```diff
@@ -0,0 +1,233 @@
+    sources_info: dict[str, QtWidgets.QWidget],
+    talkers: dict[str, ComicTalker],
+) -> None:
+    btn.clicked.connect(lambda: call_check_api_url(talker_id, sources_info, talkers))
+
+
+def format_internal_name(int_name: str = "") -> str:
+    # Presume talker_<name>_<nm>: talker_comicvine_cv_widget_name
+    int_name_split = int_name.split("_")
+    del int_name_split[0:3]
+    int_name_split[0] = int_name_split[0].capitalize()
+    new_name = " ".join(int_name_split)
+    return new_name
+
+
+def generate_checkbox(
+    option: settngs.Setting, layout: QtWidgets.QGridLayout
+) -> tuple[QtWidgets.QGridLayout, QtWidgets.QCheckBox]:
+    # bool equals a checkbox (QCheckBox)
+    widget = QtWidgets.QCheckBox(format_internal_name(option.internal_name))
+    # Add tooltip text
+    widget.setToolTip(option.help)
+    # Add widget and span all columns
+    layout.addWidget(widget, layout.rowCount() + 1, 0, 1, -1)
+
+    return layout, widget
+
+
+def generate_spinbox(
+    option: settngs.Setting, value: int | float, layout: QtWidgets.QGridLayout
+) -> tuple[QtWidgets.QGridLayout, QtWidgets.QSpinBox | QtWidgets.QDoubleSpinBox]:
+    if isinstance(value, int):
+        # int equals a spinbox (QSpinBox)
+        lbl = QtWidgets.QLabel(option.internal_name)
+        # Create a label
+        layout.addWidget(lbl, layout.rowCount() + 1, 0)
+        widget = QtWidgets.QSpinBox()
+        widget.setRange(0, 9999)
+        widget.setToolTip(option.help)
+        layout.addWidget(widget, layout.rowCount() - 1, 1, alignment=QtCore.Qt.AlignLeft)
+
+    if isinstance(value, float):
+        # float equals a spinbox (QDoubleSpinBox)
+        lbl = QtWidgets.QLabel(format_internal_name(option.internal_name))
+        # Create a label
+        layout.addWidget(lbl, layout.rowCount() + 1, 0)
+        widget = QtWidgets.QDoubleSpinBox()
+        widget.setRange(0, 9999.99)
+        widget.setToolTip(option.help)
+        layout.addWidget(widget, layout.rowCount() - 1, 1, alignment=QtCore.Qt.AlignLeft)
+
+    return layout, widget
+
+
+def generate_textbox(
+    option: settngs.Setting, layout: QtWidgets.QGridLayout
+) -> tuple[QtWidgets.QGridLayout, QtWidgets.QLineEdit, QtWidgets.QPushButton]:
+    btn = None
+    # str equals a text field (QLineEdit)
+    lbl = QtWidgets.QLabel(format_internal_name(option.internal_name))
+    # Create a label
+    layout.addWidget(lbl, layout.rowCount() + 1, 0)
+    widget = QtWidgets.QLineEdit()
+    widget.setObjectName(option.internal_name)
+    widget.setToolTip(option.help)
+    layout.addWidget(widget, layout.rowCount() - 1, 1)
+    # Special case for api_key, make a test button
+    if option.internal_name.endswith("api_key"):
+        btn = QtWidgets.QPushButton("Test Key")
+        layout.addWidget(btn, layout.rowCount() - 1, 2)
+
+    if option.internal_name.endswith("url"):
+        btn = QtWidgets.QPushButton("Test URL")
+        layout.addWidget(btn, layout.rowCount() - 1, 2)
+
+    return layout, widget, btn
```

Comment:
Why is the layout returned?

Reply:
Brain fart. I was thinking the `layout` var would be local to the function so it would need to be passed back.